### PR TITLE
feat: proposal live update progressbar

### LINKED ISSUE #460: [ProcessBar] live update elapsed time for processor bar
The elapsed time is never changed, just stuck at `3s`. 
I was wondering whether we can do a live update for elapsed time.

example code:

```go
func main() {
	showcases := []string{
		"Run short time job 1",
		"Run short time job 2",
		"Run short time job 3",
		"Run long time job 4",
	}

	pb, _ := pterm.DefaultProgressbar.
		WithTotal(len(showcases)).
		WithWriter(os.Stdout).
		Start()
	for i, showcase := range showcases {
		pb.UpdateTitle(showcase)
		if i >= 3 {
			time.Sleep(5 * time.Second)
		} else {
			time.Sleep(1 * time.Second)
		}
		pterm.Success.WithWriter(os.Stdout).Println(showcase)
		pb.Increment()
	}
}
```

git record:

pb0

### PR DESCRIPTION
**Pull request by KarolosLykos**
### Description
I thought it would be great to have real time updates on the `progressbarPrinter`. 

### Scope
Until, the new concurrency pattern is ready #439 I used the same one with `spinnerPrinter`.  These two printers are very similar and the probably will be refactored to prevent data races #482 in the future.

- [ ] Bug Fix
- [x] New Feature
- [ ] Documentation
- [ ] Other

### Related Issue
Fixes #460 

### To-Do Checklist
- [x] I tested my changes
- [x] I have commented every method that I created/changed
- [ ] I updated the examples to fit with my changes
- [x] I have added tests for my newly created methods

### CODE REVIEW

**Comment by KarolosLykos** [comment]
@MarvinJWendt what do you think is that something that would make sense?

**Comment by MarvinJWendt** [comment]
Hi, I am not quite sure if I understand your intent correctly. Why would we need live updates? Currently, the progress bar should be updated, as soon as something changes. With a fixed delay for updates, we would re-render the progress bar, even if nothing changed.
Or is this for the elapsed time?

**Comment by KarolosLykos** [comment]
> Hi, I am not quite sure if I understand your intent correctly. Why would we need live updates? Currently, the progress bar should be updated, as soon as something changes. With a fixed delay for updates, we would re-render the progress bar, even if nothing changed. Or is this for the elapsed time?

This is for the elapsed time. I saw the issue #460 and I thought I would give it a try.

**Comment by MarvinJWendt** [comment]
Ah, my bad. Yes, this would be a way to fix that. 
But we should only update on a fixed time interval when `ShowElapsedTime` is true, so that progress bars that don't show the time, aren't updated unnecessarily. Can you change that?

File: progressbar_printer.go [hunk 1]
```diff
@@ -61,101 +63,107 @@ type ProgressbarPrinter struct {
...
 	p.Title = name
-	return &p
+	return p
 }
 
 // WithMaxWidth sets the maximum width of the ProgressbarPrinter.
 // If the terminal is smaller than the given width, the terminal width will be used instead.
 // If the width is set to zero, or below, the terminal width will be used.
-func (p ProgressbarPrinter) WithMaxWidth(maxWidth int) *ProgressbarPrinter {
+func (p *ProgressbarPrinter) WithMaxWidth(maxWidth int) *ProgressbarPrinter {
```
Comment (MarvinJWendt) [inline progressbar_printer.go:74]: We may not use a pointer receiver. That way, you would overwrite the values of the "forked" printer.
For example: `pb, _ := pterm.DefaultProgressbar.WithMaxWidth(...)`, would change the `MaxWidth` of the `DefaultProgressbar` then. All `WithXxx` methods must use a non-pointer receiver.

**Comment by KarolosLykos** [comment]
> Ah, my bad. Yes, this would be a way to fix that. But we should only update on a fixed time interval when `ShowElapsedTime` is true, so that progress bars that don't show the time, aren't updated unnecessarily. Can you change that?

Yes, that make sense, I added a flag so when the `ShowElapsedTime` is off the progress bar is printed only once. When, the counter is updated the progress bar is updated.

**Comment by howieyuen** [comment]
I have clone the feature branch and test with example code in issue #460, and I found the elapsed time is accumulated from the first task, shall we print elapse time for each job ?

pb

**Comment by KarolosLykos** [comment]
@howieyuen The progress bar is meant to give an overview of the overall progress and elapsed time of all the tasks combined. It's not necessary to print the elapsed time for each individual job since it may not be very informative and can clutter the output. So, let's stick to the original  progress bar as a high-level overview of the job progress.

**Comment by howieyuen** [comment]
> @howieyuen The progress bar is meant to give an overview of the overall progress and elapsed time of all the tasks combined. It's not necessary to print the elapsed time for each individual job since it may not be very informative and can clutter the output. So, let's stick to the original progress bar as a high-level overview of the job progress.

ok, thanks for your reply

**Comment by KarolosLykos** [comment]
@MarvinJWendt Can you review and let me know your thoughts on the changes?

**Comment by MarvinJWendt** [comment]
Hi @KarolosLykos, I tried the implementation in IntelliJ (Goland) and the default Windows Terminal.

### Current implementation


### This PR


#### Windows Terminal


It seems like the terminals cannot handle so many terminal updates easily and the progress bar starts flickering.

I also tried this PR with a delay of 1 second, but that (somehow) made the flickering even worse.

This is definitely a problem with the terminals and not with this PR, but we have to consider that PTerm must work on most terminals.

Do you have any idea, how this could be fixed?

**Comment by KarolosLykos** [comment]
> Hi @KarolosLykos, I tried the implementation in IntelliJ (Goland) and the default Windows Terminal.
> 
> ### Current implementation
>  [  ](https://user-images.githubusercontent.com/31022056/232242764-1b450119-d520-4f9d-9ea5-920cb3205e9e.gif) [ ](https://user-images.githubusercontent.com/31022056/232242764-1b450119-d520-4f9d-9ea5-920cb3205e9e.gif)
> 
> ### This PR
>  [  ](https://user-images.githubusercontent.com/31022056/232243175-5d08f262-cc1a-4593-8e98-f1ae238cd55a.gif) [ ](https://user-images.githubusercontent.com/31022056/232243175-5d08f262-cc1a-4593-8e98-f1ae238cd55a.gif)
> 
> #### Windows Terminal
>  [  ](https://user-images.githubusercontent.com/31022056/232244610-022a3b85-2ed1-42d4-b1bb-fe5eeb5d1369.gif) [ ](https://user-images.githubusercontent.com/31022056/232244610-022a3b85-2ed1-42d4-b1bb-fe5eeb5d1369.gif)
> 
> It seems like the terminals cannot handle so many terminal updates easily and the progress bar starts flickering.
> 
> I also tried this PR with a delay of 1 second, but that (somehow) made the flickering even worse.
> 
> This is definitely a problem with the terminals and not with this PR, but we have to consider that PTerm must work on most terminals.
> 
> Do you have an idea, how this could be fixed?

Sorry for the late response, I understand your point of view. However, in my opinion, flickering may be unavoidable, especially when the terminal emulator's latency is high. Also, since we are testing with dummy data and not actual time-consuming tasks, the progress bar gets updated frequently, leading to more flickering.  An improved approach to handling things like this, could be a push-based solution instead of fixed interval polling.

**Comment by howieyuen** [comment]
any update on this ?

**Comment by MarvinJWendt** [comment]
Hi, this is now fixed in #500. I don't know why this PR caused flickering, but in #500 it works consistently. Maybe it's because of the other improvements in #500.

It seems, that the flickering was not caused by frequent updates, but by something else. I have tested #500 with 100 updates / second and there is no flickering at all.

Thanks to both of you @KarolosLykos and @howieyuen for helping with this!

@KarolosLykos: I will give you credits in the release, as a compensation for closing this PR

**Pull request closed without merging**